src/app/WriteClient.cpp

```diff
@@ -247,29 +248,113 @@ CHIP_ERROR WriteClient::PutSinglePreencodedAttributeWritePayload(const chip::app
     return err;
 }
 
-CHIP_ERROR WriteClient::PutPreencodedAttribute(const ConcreteDataAttributePath & attributePath, const TLV::TLVReader & data)
+CHIP_ERROR
+WriteClient::TryPutPreencodedAttributeWritePayloadIntoList(const chip::app::ConcreteDataAttributePath & attributePath,
+                                                           TLV::TLVReader & valueReader, bool & outChunkingNeeded,
+                                                           ListIndex & outEncodedItemCount)
 {
-    ReturnErrorOnFailure(EnsureMessage());
 
+    ReturnErrorOnFailure(EnsureListStarted(attributePath));
+
+    chip::TLV::TLVWriter * attributeDataIBWriter = nullptr;
+    VerifyOrReturnError((attributeDataIBWriter = GetAttributeDataIBTLVWriter()) != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    AttributeDataIB::Builder & attributeDataIB = mWriteRequestBuilder.GetWriteRequests().GetAttributeDataIBBuilder();
+
+    TLV::TLVWriter backupWriter;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    outEncodedItemCount = 0;
+
+    while ((err = valueReader.Next()) == CHIP_NO_ERROR)
+    {
+        // We essentially write List items to the Data Container within the first AttributeDataIB. We checkpoint the AttributeDataIB
+        // Builder in case there's not enough space to encode a list item.
+        attributeDataIB.Checkpoint(backupWriter);
+        err = attributeDataIBWriter->CopyElement(TLV::AnonymousTag(), valueReader);
+
+        if (err == CHIP_ERROR_NO_MEMORY || err == CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            attributeDataIB.Rollback(backupWriter);
```

Comment:
I would add `// Rollback through attributeDataIB, which will also reset the Builder error state\n // This moves the object back to as if a copy element was never attempted` or something of that form